tests/serve/serve.go

```diff
@@ -30,11 +34,27 @@ func sendThatFile(basepath string) func(w http.ResponseWriter, r *http.Request)
 
 func main() {
 	args := os.Args
-	if len(args) < 3 {
-		log.Fatal("requires listening port and subdirectory path")
+	if len(args) < 2 {
+		log.Fatal("requires subdirectory path [and optional port [and optional port file name]]")
+	}
+	basedir := args[1]
+	port := "0"
+	if len(args) > 2 {
+		port = args[2]
 	}
-	port := args[1]
-	basedir := args[2]
 	http.HandleFunc("/", sendThatFile(basedir))
-	log.Fatal(http.ListenAndServe(":"+port, nil))
+	server := http.Server{
+		Addr: ":" + port,
+		BaseContext: func(l net.Listener) context.Context {
+			if tcp, ok := l.Addr().(*net.TCPAddr); ok {
+				if len(args) > 3 {
+					if err := ioutil.WriteFile(args[3], []byte(strconv.Itoa(tcp.Port)), 0644); err != nil {
```

Comment:
This introduces a new (possibly unlikely) race condition:
* this process creates the file, starts writing port "12345" to it
* bash process sees the file is created and nonempty, reads it, gets "123" because the golang process hasn't finished writing

For as long as I can remember, I've been a fan of creating a `foo.tmp.PID` file, writing to it, closing/flushing it, then atomically renaming it to `foo`. (It's possible that `ioutil.WriteFile()` already does this, since it's such a standard pattern. In my reading of the documentation I see no indication of such).

This may be too trivial a risk to worry about in this case. I'll leave it to your discretion.

Reply:
Replaced `ioutil.WriteFile()` with `ioutils.AtomicWriteFile()`, which does the write-rename sequence.